src/stream_management/mod_stream_management.erl

```diff
@@ -386,14 +386,28 @@ maybe_handle_stream_mgmt_reroute(Acc, _StateData, _HostType, _Reason, {error, no
 handle_user_terminate(#sm_state{counter_in = H} = SmState, StateData, HostType) ->
     Sid = mongoose_c2s:get_sid(StateData),
     do_remove_smid(HostType, Sid, H),
-    reroute_buffer(StateData, SmState),
+    FromServer = mongoose_c2s:get_lserver(StateData),
+    NewState = add_delay_elements_to_buffer(SmState, FromServer),
+    reroute_buffer(StateData, NewState),
     SmState#sm_state{buffer = [], buffer_size = 0}.
 
 reroute_buffer(StateData, #sm_state{buffer = Buffer, peer = {gen_statem, {Pid, _}}}) ->
     mongoose_c2s:reroute_buffer_to_pid(StateData, Pid, Buffer);
 reroute_buffer(StateData, #sm_state{buffer = Buffer}) ->
     mongoose_c2s:reroute_buffer(StateData, Buffer).
 
+add_delay_elements_to_buffer(#sm_state{buffer = Buffer} = SmState, FromServer) ->
+    BufferWithDelays = [begin
+                            TS = mongoose_acc:timestamp(Acc),
+                            StanzaName = mongoose_acc:stanza_name(Acc),
+                            StanzaType = mongoose_acc:stanza_type(Acc),
+                            {From, To, El} = mongoose_acc:packet(Acc),
+                            ElWithDelay = maybe_add_timestamp(El, StanzaName, StanzaType, TS, FromServer),
+                            AccParams = #{from_jid => From, to_jid => To, element => ElWithDelay},
+                            mongoose_acc:update_stanza(AccParams, Acc)
+                        end || Acc <- Buffer],
+    SmState#sm_state{buffer = BufferWithDelays}.
```

Comment:
This block is doing sufficient logic to deserve its own function, so that this does instead
```
    BufferWithDelays = [ instert_delays(Acc) || Acc <- Buffer],
    SmState#sm_state{buffer = BufferWithDelays}.
```
Provided whatever is a good name for the helper.

Reply:
Yeah, I even had a similar version of this part, which also simplified the other callsite of `maybe_add_timestamp`, but in the end it was confusing. I have an idea now on how to do it better, I'll try to rework it.

Reply:
Updated, please see if it's better now. I'm not a fan of keeping functions with the same name and really different arities / expected args, but I didn't have an idea for a good name, and it should be clear which does what.

Reply:
Yeah, overloading on the arity is not my favourite thing either unless it is just the very common pattern of something like
```
do_something(Input) ->
    do_something(Input, 3).

do_something(_, 0) ->
    error;
do_something(Input, Retries) ->
    %% Some case-of logic and the usual recursion.
```

Nevertheless, it's fiiiine, not that big deal 😄